### ISSUE TITLE
Reducing verbosity of linalg::cuda::prod_slow_kernel

### ISSUE THREAD

**Issue by singam-sanjay** [viennacl/viennacl-dev#240]
Would replacing,
https://github.com/viennacl/viennacl-dev/blob/eb76eecbe4bf0bc5abf9649dd3910d56df04e35a/viennacl/linalg/cuda/matrix_operations.hpp#L1136-L1887

with launching a kernel from a 5 dimensional array in binary space filled with function pointers to CUDA kernels, with each dimension corresponding to one of the five checks made in every one of these 32 boolean expressions, e.g. , 
```

matrix_matrix_prod_kernels[row_major][row_major_A][row_major_B][transposed_A][transposed_B]<<<grid, threads>>> 
         (converted_alpha, 
           viennacl::cuda_arg(A), 
static_cast<unsigned int>(viennacl::traits::start1(A)), .... )
```

improve readability and efforts at code maintenance ?

**Comment by karlrupp** [comment]
yes. That code has just grown into its current ugly shape over time.
